internal/pkg/template/diff/write.go

```diff
@@ -0,0 +1,105 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package diff
+
+import (
+	"io"
+)
+
+const (
+	prefixAdd = "+"
+	prefixDel = "-"
+	prefixMod = "~"
+)
+
+const indentInc = 4
+
+// Writer writes the string representation of a diff tree.
+type Writer struct {
+	tree   Tree
+	writer io.Writer
+}
```

Comment:
I thought of `Tree` as the data storage structure, and how to render the data is a different responsibility. However, it could on the other hand be argued that rendering is also a part of the spec that the tree has to provide. I feel like there isn't a clear winner which is better, and I don't have a super concrete reason for going the way I chose. The real thought process was that when I realized I'd need to pass `w io.Writer` around in all these private functions, I thought maybe it's a sign that it should have been a struct on its own 😂

Reply:
oh it's fine, I was just trying to think if we can reduce the API surface of the package. We can still have a struct like this:
```

func (t *Tree) Write(w io.Writer) error {
  tw := &treeWriter{t, w}
  return tw.Write()
}

type treeWriter struct {
   tree *Tree
   out   io.Writer
}
```